### PR TITLE
Fixes #21037 - use Host::Base as auditable_type

### PR DESCRIPTION
**Pull request by iNecas**
Otherwise, we are hitting "Invalid single-table inheritance type:
Host::Discovered is not a subclass of Host::Managed" when converting the
host from discovered to managed.

### CODE REVIEW

**Comment by iNecas** [comment]
Discovery tests running here http://ci.theforeman.org/job/test_plugin_matrix/3768/

**Comment by iNecas** [comment]
@lzap this should help you with the "Invalid single-table inheritance type: Host::Discovered is not a subclass of Host::Managed" issue. The other test failures should be just regular rails5 deprecation warnings.

File: app/controllers/api/base_controller.rb [hunk 1]
```diff
@@ -67,6 +67,10 @@ def parent_scope
       association = resource_class.reflect_on_all_associations.find {|assoc| assoc.plural_name == parent_name.pluralize}
       #if couldn't find an association by name, try to find one by class
       association ||= resource_class.reflect_on_all_associations.find {|assoc| assoc.class_name == parent_name.camelize}
+      if association.nil? && parent_name == 'host'
+        association = resource_class.reflect_on_all_associations.find {|assoc| assoc.class_name == 'Host::Base'}
```
[the remark this answers is not in the document]
Comment (iNecas) [reply app/controllers/api/base_controller.rb]: whatever

**Comment by iNecas** [comment]
In case anyone interested: all green :)

**Comment by lzap** [comment]
@iNecas dude you kick ass! Lemme take a look, may I ask for double review @mhulan or anyone

**Review by lzap — APPROVED** [review]
Cool, WFM. The find vs detect was not related, wasn't it?

**Comment by iNecas** [comment]
No, it was hound barking

File: app/models/concerns/audit_extensions.rb [hunk 2]
```diff
@@ -103,8 +103,8 @@ def ensure_username
 
   def fix_auditable_type
     # STI Host class should use the stub module instead of Host::Base
-    self.auditable_type = "Host"          if auditable_type =~  /Host::/
-    self.associated_type = "Host"         if associated_type =~ /Host::/
+    self.auditable_type = "Host::Base"          if auditable_type =~  /Host::/
+    self.associated_type = "Host::Base"         if associated_type =~ /Host::/
```
Comment (ares) [inline app/models/concerns/audit_extensions.rb]: this is exactly the reason why we shouldn't align code with whitespaces :-) could you please remove them?

**Review by ares — CHANGES_REQUESTED** [review]
The problem now is that any host audit is logged as change to Host/Base (see the screenshot). If for some reason we need this, we should make sure, it's presented as a Host. It should be also listed here https://github.com/theforeman/foreman/blob/develop/app/helpers/audits_helper.rb#L2, https://github.com/theforeman/foreman/blob/develop/app/helpers/audits_helper.rb#L118 and some other places in audits helper.

There are other places that would need to be updated, https://github.com/theforeman/foreman/blob/develop/app/views/audits/show.html.erb#L6

Would it be better to take the same approach as we did for templates STI? See https://github.com/theforeman/foreman/pull/4778 for what I mean.

host

**Comment by iNecas** [comment]
Screenshot missing

**Comment by iNecas** [comment]
I don't think https://github.com/theforeman/foreman/pull/4778 is the same, as in host case we actually change the type of the class as we go (from Discovered to Managed), while in templates, this is immutable. I will try to update the missed places.

**Comment by iNecas** [comment]
Nm for the screenshots: I've seen it by myself: updated + addressed other comments.

**Comment by lzap** [comment]
LGTM @ares ?

**Comment by ares** [comment]
Sorry to keep asking... how does this happen in the first place? I don't see `audited` definition in Host::Discovered nor Host::Base, it's in Host::Managed only. AFAIK there's no Managed -> Discovered transition. I don't want to block it, just want to understand, what we're trying to solve here. This basically means any subclass of Host will be audited as generic Host. No way to search audits based on host type. I know we have only managed and discovered hosts so far, but before we commit to this, I'd like to know why :-)

**Comment by iNecas** [comment]
That's definitely valid reason to ask.

The problem is, when saving the audit records for the host, the audit record is being created first, and as part of the process, the auditable record (the host) is being loaded again from the database: and this is where the issue comes, as the auditable_type was marked as `Host` which translates to `Host::Managed` and the `Host::Discovered` was set in the database, which lead to the error `"Invalid single-table inheritance type: Host::Discovered is not a subclass of Host::Managed`.

Since we allow transitions of types for host record, I think it makes sense to keep the auditable_type as `Host::Base`. Honestly, the Host should not be an STI in first place and discovered host should be just a record on it's own: but that's outside of scope for this PR

**Comment by lzap** [comment]
@ares hey are we good to go?

File: app/helpers/audits_helper.rb [hunk 3]
```diff
@@ -115,7 +115,7 @@ def audited_icon(audit)
 
     type   = audited_type(audit)
     symbol = case type
-               when "Host"
+               when "Host::Base"
```
Comment (ares) [inline app/helpers/audits_helper.rb]: this is `'Host'` actually, since audited_type converts it

**Review by ares — CHANGES_REQUESTED** [review]
Just one tiny thing re icons, otherwise seems to work fine, good to merge afterwards

**Review by ares — APPROVED** [review]

**Comment by ares** [comment]
Ready for merge pending tests

**Comment by iNecas** [comment]
Tests are green

**Comment by lzap** [comment]
Thanks a bunch folks, really appreciated for unblocking discovery tests.

**Comment by ares** [comment]
I set the release to 1.17, @lzap if you want this to be cherry-picked to 1.16, please updated the redmine issue. If it's accepted, then you should also do the cherry-pick to 1.16-stable.

**Comment by kotyara85** [comment]
Facing the same issue on 1.17. Can't provision hosts unless I'm an admin. https://projects.theforeman.org/issues/23869

**Pull request merged**